src/cast.rs

```diff
@@ -164,3 +159,71 @@ where
         }
     }
 }
+
+/// Branch on the instance type of a Java object. It will execute (and only execute) the first match arm that has a type
```

Comment:
Hmm, I feel like, rather than going the macro route, we could instead do something with methods. I saw your comment about the subtle difference of chained `catch` methods, but that assumes we do a naive chaining.

We could instead do something like this...

```rust
jvm_op
    .catch_block()
    .catch(|x: Bar| ...)
```

where `catch_block` is a method on `JniOp` and `catch(self)` is an (inherent) method on the `Catch` operator type itself. The idea would be that `catch_next` doesn't apply to exceptions that occur inside of `catch_first`, and instead adds arms into the catch block. We probably want a  `catch_otherwise` or `finally` as well. We could have a complete method to exit the catch block if we wanted, or simply have it also implement `JniOp`.

Another option might be...

```rust
x.catch_first(....).catch_next(...).finally(...)
```

you could even do `x.catch().catch()` and rely on the Rust method handle (which gives precedence to inherent methods), but that seems a bit subtle and error prone for my taste.

Reply:
You're right, I think I'm conflating a couple of issues here.

We can fix the chained catches with something like `catch_block()` or `catch_one_of()`. I'l go ahead and impl that since it's cleaner than the macro for that case. 

That said, I do expect folks to need to inspect and branch on the type of a returned object, be it an exception or regular data. In the auth example, I imagine the real code looking like

```rust
let auth_result: Result<User, AuthFailure> = Jvm::with(|jvm| {
    let http_request = HttpRequest::new(/* ... */);
    match authenticator.authenticate(http_request).execute(jvm) {
        Ok(authenticated_user) => Ok(Ok(build_user(authenticated_user))),
        Err(duchess::Error::Thrown(t)) => by_type! {
            with jvm match t => {
                InvalidSignatureException as t => Ok(Err(AuthFailure::InvalidSignature(t.get_message().execute(jvm)?))),
                InternalError => Ok(Err(AuthFailure::InternalError)),
                // ...
            }
        }
        Err(e) => Err(e),
}).unwrap();
```

I don't think trying to put that "into Rust" logic in the catch JvmOp will work since they're not types that can be put into Java land.

Reply:
Added the `catching()` fn in the next commit

Reply:
Hmm, let's talk this one out.